ReadME.md
[Shopping Website](https://howardshou.github.io/real-life-project/)
===
This is my first project while I'm working on becoming a qualified F2E engineer, mainly focus on making a shopping website demo, not only includes a product list, detail, shopping cart, checkout form but also contains different layout practices like blogs, which makes me have a solid foundation for CSS and HTML.

Please click title or [here](https://howardshou.github.io/real-life-project/) to visit the website.




changelog.md
9/14

於index的修正layout部分設定--將原本某些設定content-box的部分修正為border-box

雖然用bootstrap的grid系統刻出來，就width的控制沒有問題，不過每張圖片容器高度本身是用所給的圖片撐出高度的
因此只要圖片的高度不同，容器的大小高低也會因此有差異

如果實務上會要求「以任何大小的圖片塞入容器皆不影響高度」，針對img的tag，我除了原本設定的display:block;與width:100%;
外，我會另外增加height:100%;，而裝有該tag的主容器我會給予固定的height，應可達到要求

-----------------------------------------

9/2

修正上次最後檢討的小錯誤後，完成githubpage版本

-----------------------------------------

8/26

目前看起來應該只有form的select元素展開時有不正常顯示的現象
看過MDN等資料是說select的內部架構很複雜，不建議自己把appearance清掉自己style
本次基於嘗試還是自己清掉來實作，如果真的有問題會再引進jQuery UI並style

-------------------------------------------

8/23

Q 到底要如何用同樣的架構達成「991到768一行四排，767以下一行兩排」

上述問題在product_list的html,code 289以下，搭配less code 1047以下要實作重現，看有沒有解決方式

------------------------------------------

8/22

1.blog_post的content尾端的comments，雖然看起來版面是沒有問題，不過總覺得該部分的html的架構可能有問題...

2.product_detail關於input[type="number"]，找過資料後，目前還是無法弄出客製化的spinner...

-------------------------------------------


8/16

1.最大的問題還是在FDI_1與FDI_2這兩個大容器的切版，很明顯還是跟設計稿有落差，但是我真的盡力了...
應該是我哪邊的架構沒有規劃好

2.呈上，可以看的出來，雖然沒有出現scrollbar，但是FDI這兩個容器還是多出了不必要的白邊，我研究了兩小時以上
...DEV-toll看到快爛掉了，也嘗試改html架構，仍舊找不出問題在哪，希望能找出盲點T_T

3.覆蓋、置中、absolute要設在哪個點上比較適合等，是我現在在做RWD時最常遇到的狀況，雖然最後大都能解決問題
，不過總是要嘗試各種方法才會知道最佳解，除了多練習外，之後有機會也想參考看看助教和老師在面對這些設計稿時
是怎麼切版，對齊、覆蓋等技巧，相信自己這樣一直摸索下來，在觀摩資深的你們下的邏輯架構，應該會有些體悟XD

-------------------------------------------
8/12

1.了解margin:0 -15px設置的意義，解決跑版問題

2.順便修正在xs small device(width<=767px)時nav靠太近與一兩個小地方跑版




-------------------------------------------
8/11

1.刪除非必要的外掛，能用bootstrap就用bootstrap

2.兩個問題:

2.1 bootstrap navbar的使用問題:
由於設計稿的nav很長，因此畫面一縮小很快就會產生推擠，雖然可以用隱藏元素的方式解決
但是應該還是要開一個off-canvas之類的讓使用者可以點選吧?是否有其他更好的解決方案呢...

2.2 找不出的留白:
觀察後，發現當視窗約小於900px時，右邊似乎有殘留的padding導致畫面右邊總會有一點留白，
視窗越小越明顯，我找了好久還是找不出問題所在QQ... 


js/jquery.responsive-nav.js
/**
 * jquery.responsive-nav
 * Description: レスポンシブなナビゲーションを実装。プルダウンナビ <=> オフキャンバスナビ
 * Version: 1.1.3
 * Author: Takashi Kitajima
 * Autho URI: http://2inc.org
 * created : February 20, 2014
 * modified: February 5, 2015
 * package: jquery
 * License: GPLv2
 * License URI: http://www.gnu.org/licenses/gpl-2.0.html
 */

;( function( $ ) {
		$.fn.responsive_nav = function( config ) {
			var is_open     = false;
			var container   = $( '#responsive-nav-wrapper' );
			var responsive_nav = this;
			var offcanvas_nav;

			return this.each( function( i, e ) {
				offcanvas_nav = responsive_nav.clone( true );
				responsive_nav.addClass( 'responsive-nav' );
				init();

				var menu = $( this ).find( 'ul:first-child' );
				menu.children( 'li' ).children( 'ul' ).children( 'li' ).hover( function() {
					var children = $( this ).children( 'ul' );
					if ( children.length ) {
						if ( $( window ).width() < ( children.width() + children.offset().left ) ) {
							children.addClass( 'reverse-pulldown' );
							children.find( 'ul' ).addClass( 'reverse-pulldown' );
						}
					}
				} );

				$( window ).resize( function() {
					var children = menu.children( 'li' ).children( 'ul' ).children( 'li' );
					children.removeClass( 'reverse-pulldown' );
					children.find( 'ul' ).removeClass( 'reverse-pulldown' );
					init();
					if ( is_open ) {
						nav_close();
					}
				} );

				$( '#responsive-btn' ).click( function() {
					if ( is_open ) {
						nav_close();
					} else {
						nav_open();
					}
				} );
				$( 'html' ).click( function() {
					if ( is_open ) {
						nav_close();
					}
				} );
			} );

			function init() {
				if ( $( '#responsive-btn' ).css( 'display' ) !== 'none' ) {
					wrap_all();
				} else {
					unwrap();
					unset_off_canvas_nav();
				}
			}

			function get_slide_width() {
				return offcanvas_nav.width();
			}

			function nav_open() {
				var height = get_window_height();
				offcanvas_nav.css( {
					height: height
				} );
				offcanvas_nav.show();
				container.css( 'width', container.width() );
				container.animate( {
					marginLeft: get_slide_width()
				}, 200, function() {
					is_open = true;
					$( 'body' ).addClass( 'nav-open' );
				} );
			}

			function nav_close() {
				container.animate( {
					marginLeft: 0
				}, 200, function() {
					unwrap();
					wrap_all();
				} );
			}

			function unwrap() {
				is_open = false;
				container.css( {
					'width'     : '',
					'marginLeft': ''
				} );
				$( 'body' ).removeClass( 'nav-open' );
				responsive_nav.show();
				offcanvas_nav.remove();
			}

			function wrap_all() {
				container.prepend( offcanvas_nav );
				responsive_nav.hide();
				offcanvas_nav.addClass( 'off-canvas-nav' );
				offcanvas_nav.css( 'left', - get_slide_width() );
			}

			function get_window_height() {
				var height      = $( window ).height();
				var html_margin = parseInt( $( 'html' ).css( 'marginTop' ) );
				height = height - html_margin;
				return height;
			}

			function unset_off_canvas_nav() {
				offcanvas_nav.removeClass( 'off-canvas-nav' );
				offcanvas_nav.css( 'display', 'block' );
			}
		}
} )( jQuery );

// jQuery( function( $ ) {
//   $('#navi').responsive_nav();
// });



plug-in&筆記/notes/main._note.js
$(document).ready(function() {
  
  /*自己寫出來了，給自己牡蠣!，各種問題包含要使用on、用什麼方法等等
  其中有一個關鍵，就是測試對jq時，對照著開法者模式的elemtents看!
  像我後來發現很多時候一開始jq會動，後來卻不理我的關鍵，就是在選擇器下錯了!
  會下錯的原因在於，jq啟用時常變動了html的架構，既然架構已經變動了，那就
  要照著變動過後的架構去設定選擇器，否則它當然會不知道你在跟它說啥!*/

  /*Q1原本想要在全域設定以下變數，裝jQ物件，以精簡原本的程式碼，
    但是照註解的部分這樣設定瀏覽器看不懂，我錯在哪...*/
  /*var Nav_Others = $('#nav_others').clone();
    var Cart = $('#cart-container');
    var Showing = $('#cart-container.cartshow');
    var Hiding = $('#cart-container.carthide');*/

  // 一定要用on，因為你所更動、設定的物件，一開始是不存在的
  $('body').on('click', '#responsive-btn-cart', function() {
    var nav_others = $('#nav_others').clone();
    // 這邊是用on的原因--.carthide一開始並未存在於html中
    $('#cart-container.carthide').removeClass('carthide')
    /* Q2:為何不會每點一次'#responsive-btn-cart'就插入一串ul導致容器塞爆?
       A2:html這個method並不是像before和after那些method一樣是單純的插入
         而是「先清空在插入」，是一個「覆蓋」的概念!*/ 
    $('#cart-container').html(nav_others);
    $('#cart-container').addClass('cartshow')  
  });
  
  // 上面在#cart-container新添加了.cartshow，既然它為後來才加入html的class，必須用on
  /* Q3:我本來想要製造效果:當off-cnavas顯現中，我點body的任一處，就可以讓它自動收合
        所以想這樣寫:$('body').on('click', 'body', function() {......}，但失敗了
        ，下面這個寫法雖然能順利地將off-canvas關起來，但是僅限點擊黑色範圍內，
        而不同於我想要的效果:典籍畫面內任一處即使之消失*/

  $('body').on('click', '#cart-container.cartshow', function() {
    // 如此一來能正確吃到css動畫效果:讓width從33%縮回0%
    $('#cart-container.cartshow').removeClass('cartshow').addClass('carthide')
    // 把width縮回0後，由於我們之前clone進來的nav_others還在，必須要將他的內容清空，才不會留在畫面中
    $('#cart-container.carthide').empty()
    // $('#cart-container.carthide').css('display', 'none');
    /*設定dsiplay:none有兩個問題，
    1.在carthide被設定過的css不會在show時自動消失，代表我們需要對$('#cart-container.cartshow')
      設定display:block
    2.會導致畫面瞬間消失，而不是我們想要的動畫效果
      */
  });


  $(window).resize(function(){
    if ($(window).width() >= 768){
      $('#cart-container').removeClass();
      /* 如果是平板以上的視窗/容器，則讓cart-container回歸到初始狀態，如果是用css('display','none')
         則還得再多寫一行 else{...}，會比較沒效率
         $('#cart-container').css('display', 'none');
         } 
         else{
         $('#cart-container').css('display', 'block');
         */   
    }
  });

  $('#navi').responsive_nav();

  /*失敗的寫法
  $('.photos').carousel({
    num: 3, //要显示的数量，应该是个奇数
    maxWidth: 1197, //代表中央图片的宽度
    maxHeight: 542, //中央图片的高度
    autoPlay: true, //是否自动滚动播放
    showTime: 4000, //autoPlay为true时这个属性才有用
    animationTime: 1000, //
    scale: 0.8,
    distance: 50);
  });*/

  // 練習插入他人的外掛:呼叫可以折疊文字的外掛
  var myAccordion = new $.Zebra_Accordion('.Zebra_Accordion',{
    collapsible:true,
    show:true
  })


});



/*如同我自己的體驗，和老師所說，修別人的東西的確會比較痛苦T_T
  我只是想要捯到一個效果:在視窗於平板調腦以上時，維持原狀就好，等到視窗是為平板以下，塞入插建
  然後在視窗又回到平板以上時，回復原狀

  比想像中的難超多...!尤其是怎麼在變動視窗時停用外掛這件事上，如果不是單純用+-class的方法，
  外掛一被插入，更動了css，甚至html，常常就一去不復返了，關不掉...

  而且這個外掛還有一個地方很怪，又或許是我jQ寫錯...
  如果是後來才插入這個外掛，外掛會失靈，不會給dd height !導致dd裡的元素會跟下面的元素疊在一起...
  我已經花太多時間，既然自始就插入這個只有2kb的外掛對效能來說應該問題不大，就放著吧!

  以下是其他嘗試失敗的設定，想要利用+-calss去控制等等
  $('#Footer .col-5 .col-4 dl.Zebra_Accordion').removeClass('Zebra_Accordion');
  $('#Footer .col-5 .col-4 dl dd').css({
    display: 'block',
    opacity: '1',
    height:'auto'
  });
*/



   

plug-in&筆記/notes/Carousel._note1.js
/* Q1:助教有說，盡量不要直接改外掛的程式碼，而是另外自己在寫一個去搭配、修正，如此將來方便維護
    不過...自己寫相對困難...而且這些外掛的語法還參雜著JS改的真的很痛苦...是否還是先在外掛裡面改
    比較好?
    ps以下的問題我都有自己上網查過，但可能是關鍵字下的不對...或者是回答的答案我看不懂T_T
    我覺得應該有不少是我對jQ的語法不夠理解...每次查資料卻還是看不懂，覺得挫折...*/

/*  Q2:這個外掛中，它原本預設的選擇器是$container，我知道這個寫法應該代表著var $container = $('.container')之類的
      ，我想要它變成var $container = $(.photos)這樣，我試著找出$container到底指哪個選擇器，但我找破頭還是 找不到作者
      在哪裡有宣告var $container到底裝什麼!!!!真的很崩潰!!!
        
/*  Q 3.1:我自己最後想到的方法是...既然找不到，我自己宣告看看...先在全域設一個var $container = $('.photos') (line:57)
          但又不確定有沒有用，索性在每一支函式函式都宣告一次(line:72、92等)，終於更動了作者的設定，但我發現仍就有瑕疵...雖然
           效果有出來，但是position設的位置會跑掉(line:266)...所以還是有bug...
      
      3.2 另外，我還有做一個嘗試，就是直接把非參數的$container全部直接用$('.photos')替換，結果還是失敗了...想知道為什麼
          不能直接這樣替代...
      */

/*  Q4:我看很多外掛起手式固定寫法:(function ($){--blablabla---}(jQuery); 參數為$，用(jQuery)結尾，中間塞所有的變數、函式等
      現在jQ3.3.1版本還是這樣寫嗎?它的意義是什麼?*/
 
 
// 以下為本carousel的js的程式碼
// 以下為本carousel的js的程式碼
// 以下為本carousel的js的程式碼
(function ($) {
    if ($ === undefined) {
        return;
    }
  
    // var $container = $('.photos')

    var defaultConfig = {
        num: 3, //要显示的数量，应该是个奇数
        maxWidth: 250, //代表中央图片的宽度
        maxHeight: 150, //中央图片的高度
        autoPlay: true, //是否自动滚动播放
        showTime: 1000, //autoPlay为true时这个属性才有用
        animationTime: 300, //
        scale: 0.8,
        distance: 50
    };


    function getzIndexValue(num, direction) {
      // var $container = $('.photos')
        var zIndexs = [];
        for (var i = 0; i < num; i++) {
            if (i <= (num - 1) / 2) {
                zIndexs.push(i);
            } else {
                zIndexs.push((num - 1) / 2 - i);
            }
        }
        if (direction === 'left') {
            zIndexs.reverse();
            return zIndexs;
        }
        if (direction === 'right') {
            return zIndexs;
        }

    }

    function scroll($container, direction) {
      // var $container = $('.photos')
        if ($container.data('isanimating')) {
            return;
        }
        var config = $container.data('config');
        var halfShowNum = (config.num - 1) / 2;
        var scales, i, newIndex;
        var totalNum = $container.data('totalNum');
        var targetCss;
        var firstIndexBeforeScroll, lastIndexBeforeScroll;
        if (direction === 'left') {
            newIndex = ($container.data('index') - 1 + totalNum) % totalNum;
        } else if (direction === 'right') {
            newIndex = ($container.data('index') + 1) % $container.data('totalNum');
        } else {
            return;
        }
        // $container.find('ul li').stop(true, true);
        var tempIndexsInfo = getShowIndexs($container);
        firstIndexBeforeScroll = tempIndexsInfo.indexs[0];
        lastIndexBeforeScroll = tempIndexsInfo.indexs[config.num - 1];
        $container.data('index', newIndex);
        var showIndexsInfo = getShowIndexs($container);
        var zIndexs = getzIndexValue(config.num, direction);
        if (totalNum === config.num) {
            animationTimeForEdge = 0
        } else if (totalNum - config.num === 2) {
            animationTimeForEdge = config.animationTime / 2;
        } else {
            animationTimeForEdge = config.animationTime;
        }

        /*
         showIndexsInfo = {
         indexs: [5, 6, 0, 1, 2]
         hashIndexs: {
         '5': 0,
         '6': 1,
         '0': 2,
         '1': 3,
         '2': 4
         }
         }
         */
        $container.find('ul li').each(function (index, element) {
          // var $container = $('.photos')  
            i = showIndexsInfo.hashIndexs[index];

            if (i !== undefined) {
                scales = Math.pow(config.scale, Math.abs(i - halfShowNum));
                $container.data('isanimating', true);
                $(element).css({
                    display: 'block',
                    'z-index': zIndexs[i] + 9999
                }).animate({
                    width: scales * config.maxWidth,
                    height: scales * config.maxHeight,
                    left: i * config.distance + (1 - scales) * config.maxWidth * Number(i > halfShowNum),
                    top: (1 - scales) * config.maxHeight / 2
                }, config.animationTime, function () {
                    $container.data('isanimating', false);
                });

            } else {
                scales = Math.pow(config.scale, halfShowNum);
                //if(direction === 'right' && index === firstIndexBeforeScroll){
                //    console.log('right' + index);
                //} else if(direction === 'left' && index === lastIndexBeforeScroll) {
                //    console.log('left' + index);
                //}

                targetCss = {
                    display: 'none',
                    left: halfShowNum * config.distance + (1 - scales) * config.maxWidth / 2,
                    top: 0
                };
                if (direction === 'left' && index === lastIndexBeforeScroll) {

                    $(element).css('z-index', -1).animate({
                        left: "-=" + config.distance + "px"
                    }, config.animationTime, function () {
                        $(element).css(targetCss);
                    });
                } else if (direction === 'right' && index === firstIndexBeforeScroll) {

                    $(element).css('z-index', -1).animate({
                        left: "+=" + config.distance + "px"
                    }, config.animationTime, function () {
                        $(element).css(targetCss);
                    });
                } else {
                    $(element).css({
                        display: 'none',
                        width: scales * config.maxWidth,
                        height: scales * config.maxHeight,
                        left: halfShowNum * config.distance + (1 - scales) * config.maxWidth / 2,
                        top: 0
                    });
                }
            }

        });
    }

    function getConfig(newConfig) {
      // var $container = $('.photos')
        var config = null;
        if (typeof newConfig === 'object' && newConfig !== null) {
            config = {};
            for (var prop in defaultConfig) {
                if (defaultConfig.hasOwnProperty(prop)) {
                    config[prop] = defaultConfig[prop];
                }
            }
            for (prop in newConfig) {
                if (newConfig.hasOwnProperty(prop) && config.hasOwnProperty(prop)) {
                    config[prop] = newConfig[prop];
                }
            }
        }
        return config;
    }

    function getShowIndexs($container) {
      // var $container = $('.photos')
        var showIndexs = [];
        var temp;
        var halfShowNum = ($container.data('config').num - 1) / 2;
        var currentIndex = $container.data('index') || 0;
        var totalNum = $container.data('totalNum') || 0;
        for (var i = -halfShowNum; i <= halfShowNum; i++) {
            temp = currentIndex + i;
            showIndexs.push((temp < 0 ? (temp + totalNum) : temp) % totalNum);
        }
        var hashIndexs = {};
        for (i = 0; i < showIndexs.length; i++) {
            hashIndexs[showIndexs[i]] = i;
        }
        return {
            indexs: showIndexs,
            hashIndexs: hashIndexs
        };
    }

    function initStyle($container) {
      // var $container = $('.photos')
        var showIndexsInfo = getShowIndexs($container);

        var zIndex = 9999;
        var scales;
        var config = $container.data('config');
        var halfShowNum = (config.num - 1) / 2;
        var listWidth = halfShowNum * config.distance * 2 + config.maxWidth;
        var containerWidth = $container.width();
        var containerHeight = $container.height();
        if (containerWidth < listWidth) {
            $container.width(listWidth);
        }
        if (containerHeight < config.maxHeight) {
            $container.height(config.maxHeight);
        }
        $container.find('ul li img').css({
            width: "100%",
            height: "100%"
        });
        $container.find('ul').css({
            position: 'relative',
            width: listWidth,
            height: config.maxHeight,
            'list-style': 'none',
            padding: 0,
            margin: 0,
            marginLeft: '50%',
            marginRight: '50%',
            left: -listWidth / 2,
            /*customize:
            top: ($container.height() - config.maxHeight) / 2
            */
        });

        $container.find('.left').css({
            position: 'absolute',
            left: 10,
            top: '50%',
            'z-index': 9999 + $container.data('totalNum') + 1
        });

        $container.find('.right').css({
            position: 'absolute',
            right: 10,
            top: '50%',
            'z-index': 9999 + $container.data('totalNum') + 1
        });

        $container.find('ul li').each(function (index, element) {
            var i = showIndexsInfo.hashIndexs[index];
            if (i !== undefined) {
                scales = Math.pow(config.scale, Math.abs(i - halfShowNum));
                zIndex = 9999 + (i > halfShowNum ? (config.num - 1 - i) : i);
                $(element).css({
                    display: 'block',
                    position: 'absolute',
                    'z-index': zIndex,
                    overflow: 'hidden',
                    width: scales * config.maxWidth,
                    height: scales * config.maxHeight,
                    left: i * config.distance + (1 - scales) * config.maxWidth * Number(i > halfShowNum),
                    top: (1 - scales) * config.maxHeight / 2
                });
            } else {
                scales = Math.pow(config.scale, halfShowNum);
                $(element).css({
                    display: 'none',
                    position: 'absolute',
                    overflow: 'hidden',
                    width: scales * config.maxWidth,
                    height: scales * config.maxHeight,
                    left: halfShowNum * config.distance + (1 - scales) * config.maxWidth / 2,
                    top: 0
                });
            }

        });
    }


    $.fn.carousel = function (param) {
      // var $container = $('.photos')
        var config;
        var totalNum;
        var $target;
        $(this).each(function(index, target) {
            $target = $(target);
            if (typeof param === 'object' && param !== null) {
                config = getConfig(param);
                totalNum = $target.find('ul li').length;
                if (totalNum <= 0 || totalNum % 2 === 0) {
                    return;
                }
                if (config.num <= 0 || config.num > totalNum) {
                    config.num = totalNum;
                }
                $target.data('config', config);
                $target.data('index', 0);
                $target.data('totalNum', totalNum);
                initStyle($target);

                $target.find('.left').off('click').on('click', (function($target) {
                    return function() {
                        scroll($target, 'left');
                    }
                })($target));
                $target.find('.right').off('click').on('click',(function($target) {
                    return function() {
                        scroll($target, 'right');
                    }
                })($target));

                (function($target) {
                    var autoPlay;
                    clearInterval($target.data('auto'));
                    if($target.data('config').autoPlay) {
                        autoPlay = setInterval(function() {
                            scroll($target, 'right');
                        }, $target.data('config').showTime);
                        $target.data('auto', autoPlay);
                        $target.find('ul').off('mouseenter').on('mouseenter', function() {
                            clearInterval($target.data('auto'));
                        }).off('mouseleave').on('mouseleave', function() {
                            autoPlay = setInterval(function() {
                                scroll($target, 'right');
                            }, $target.data('config').showTime);
                            $target.data('auto', autoPlay);
                        });
                    } else {
                        $target.find('ul').off('mouseenter').off('mouseleave');
                    }
                })($target);
            }

        });

    };

})(jQuery);



js/Carousel.js
/* Q1:助教有說，盡量不要直接改外掛的程式碼，而是另外自己在寫一個去搭配、修正，如此將來方便維護
    不過...自己寫相對困難...而且這些外掛的語法還參雜著JS改的真的很痛苦...是否還是先在外掛裡面改
    比較好?
    ps以下的問題我都有自己上網查過，但可能是關鍵字下的不對...或者是回答的答案我看不懂T_T
    我覺得應該有不少是我對jQ的語法不夠理解...每次查資料卻還是看不懂，覺得挫折...*/

/*  Q2:這個外掛中，它原本預設的選擇器是$container，我知道這個寫法應該代表著var $container = $('.container')之類的
      ，我想要它變成var $container = $(.photos)這樣，我試著找出$container到底指哪個選擇器，但我找破頭還是 找不到作者
      在哪裡有宣告var $container到底裝什麼!!!!真的很崩潰!!!
        
/*  Q 3.1:我自己最後想到的方法是...既然找不到，我自己宣告看看...先在全域設一個var $container = $('.photos') (line:57)
          但又不確定有沒有用，索性在每一支函式函式都宣告一次(line:72、92等)，終於更動了作者的設定，但我發現仍就有瑕疵...雖然
           效果有出來，但是position設的位置會跑掉(line:266)...所以還是有bug...
      
      3.2 另外，我還有做一個嘗試，就是直接把非參數的$container全部直接用$('.photos')替換，結果還是失敗了...想知道為什麼
          不能直接這樣替代...
      */

/*  Q4:我看很多外掛起手式固定寫法:(function ($){--blablabla---}(jQuery); 參數為$，用(jQuery)結尾，中間塞所有的變數、函式等
      現在jQ3.3.1版本還是這樣寫嗎?它的意義是什麼?*/
 

/*  Q5這個問題跟carousel無關，這是我自己寫的jQ遇到的問題:
    
    $(document).ready(function(){
    
    原本想要在全域設定宣告以下變數，指派為jQ物件，以精簡原本的程式碼，
    但是這樣宣告瀏覽器看不懂(ex line:37、42)，我錯在哪?宣告的方式錯了?不能設在全域?
    var Nav_Others = $('#nav_others').clone();
    var Cart = $('#cart-container');
    var Showing = $('#cart-container.cartshow');
    var Hiding = $('#cart-container.carthide');

    $('body').on('click', '#responsive-btn-cart', function() {
      var nav_others = $('#nav_others').clone();
      $('#cart-container.carthide').removeClass('carthide')
      $('#cart-container').html(nav_others);→本來想寫:Cart.html(nav_others);
      $('#cart-container').addClass('cartshow')  
    });
  
    $('body').on('click', '#cart-container.cartshow', function() {
      $('#cart-container.cartshow').removeClass('cartshow').addClass('carthide');→本來想寫:Showing.removeClass('cartshow').addClass('carthide')
      $('#cart-container.carthide').empty()
    });
  });
*/
    

// 以下為本carousel的js的程式碼
// 以下為本carousel的js的程式碼
// 以下為本carousel的js的程式碼
(function ($) {
    if ($ === undefined) {
        return;
    }
  
    // var $container = $('.photos')

    var defaultConfig = {
        num: 3, //要显示的数量，应该是个奇数
        maxWidth: 250, //代表中央图片的宽度
        maxHeight: 150, //中央图片的高度
        autoPlay: true, //是否自动滚动播放
        showTime: 1000, //autoPlay为true时这个属性才有用
        animationTime: 300, //
        scale: 0.8,
        distance: 50
    };


    function getzIndexValue(num, direction) {
      // var $container = $('.photos')
        var zIndexs = [];
        for (var i = 0; i < num; i++) {
            if (i <= (num - 1) / 2) {
                zIndexs.push(i);
            } else {
                zIndexs.push((num - 1) / 2 - i);
            }
        }
        if (direction === 'left') {
            zIndexs.reverse();
            return zIndexs;
        }
        if (direction === 'right') {
            return zIndexs;
        }

    }

    function scroll($container, direction) {
      // var $container = $('.photos')
        if ($container.data('isanimating')) {
            return;
        }
        var config = $container.data('config');
        var halfShowNum = (config.num - 1) / 2;
        var scales, i, newIndex;
        var totalNum = $container.data('totalNum');
        var targetCss;
        var firstIndexBeforeScroll, lastIndexBeforeScroll;
        if (direction === 'left') {
            newIndex = ($container.data('index') - 1 + totalNum) % totalNum;
        } else if (direction === 'right') {
            newIndex = ($container.data('index') + 1) % $container.data('totalNum');
        } else {
            return;
        }
        // $container.find('ul li').stop(true, true);
        var tempIndexsInfo = getShowIndexs($container);
        firstIndexBeforeScroll = tempIndexsInfo.indexs[0];
        lastIndexBeforeScroll = tempIndexsInfo.indexs[config.num - 1];
        $container.data('index', newIndex);
        var showIndexsInfo = getShowIndexs($container);
        var zIndexs = getzIndexValue(config.num, direction);
        if (totalNum === config.num) {
            animationTimeForEdge = 0
        } else if (totalNum - config.num === 2) {
            animationTimeForEdge = config.animationTime / 2;
        } else {
            animationTimeForEdge = config.animationTime;
        }

        /*
         showIndexsInfo = {
         indexs: [5, 6, 0, 1, 2]
         hashIndexs: {
         '5': 0,
         '6': 1,
         '0': 2,
         '1': 3,
         '2': 4
         }
         }
         */
        $container.find('ul li').each(function (index, element) {
          // var $container = $('.photos')  
            i = showIndexsInfo.hashIndexs[index];

            if (i !== undefined) {
                scales = Math.pow(config.scale, Math.abs(i - halfShowNum));
                $container.data('isanimating', true);
                $(element).css({
                    display: 'block',
                    'z-index': zIndexs[i] + 9999
                }).animate({
                    width: scales * config.maxWidth,
                    height: scales * config.maxHeight,
                    left: i * config.distance + (1 - scales) * config.maxWidth * Number(i > halfShowNum),
                    top: (1 - scales) * config.maxHeight / 2
                }, config.animationTime, function () {
                    $container.data('isanimating', false);
                });

            } else {
                scales = Math.pow(config.scale, halfShowNum);
                //if(direction === 'right' && index === firstIndexBeforeScroll){
                //    console.log('right' + index);
                //} else if(direction === 'left' && index === lastIndexBeforeScroll) {
                //    console.log('left' + index);
                //}

                targetCss = {
                    display: 'none',
                    left: halfShowNum * config.distance + (1 - scales) * config.maxWidth / 2,
                    top: 0
                };
                if (direction === 'left' && index === lastIndexBeforeScroll) {

                    $(element).css('z-index', -1).animate({
                        left: "-=" + config.distance + "px"
                    }, config.animationTime, function () {
                        $(element).css(targetCss);
                    });
                } else if (direction === 'right' && index === firstIndexBeforeScroll) {

                    $(element).css('z-index', -1).animate({
                        left: "+=" + config.distance + "px"
                    }, config.animationTime, function () {
                        $(element).css(targetCss);
                    });
                } else {
                    $(element).css({
                        display: 'none',
                        width: scales * config.maxWidth,
                        height: scales * config.maxHeight,
                        left: halfShowNum * config.distance + (1 - scales) * config.maxWidth / 2,
                        top: 0
                    });
                }
            }

        });
    }

    function getConfig(newConfig) {
      // var $container = $('.photos')
        var config = null;
        if (typeof newConfig === 'object' && newConfig !== null) {
            config = {};
            for (var prop in defaultConfig) {
                if (defaultConfig.hasOwnProperty(prop)) {
                    config[prop] = defaultConfig[prop];
                }
            }
            for (prop in newConfig) {
                if (newConfig.hasOwnProperty(prop) && config.hasOwnProperty(prop)) {
                    config[prop] = newConfig[prop];
                }
            }
        }
        return config;
    }

    function getShowIndexs($container) {
      // var $container = $('.photos')
        var showIndexs = [];
        var temp;
        var halfShowNum = ($container.data('config').num - 1) / 2;
        var currentIndex = $container.data('index') || 0;
        var totalNum = $container.data('totalNum') || 0;
        for (var i = -halfShowNum; i <= halfShowNum; i++) {
            temp = currentIndex + i;
            showIndexs.push((temp < 0 ? (temp + totalNum) : temp) % totalNum);
        }
        var hashIndexs = {};
        for (i = 0; i < showIndexs.length; i++) {
            hashIndexs[showIndexs[i]] = i;
        }
        return {
            indexs: showIndexs,
            hashIndexs: hashIndexs
        };
    }

    function initStyle($container) {
      // var $container = $('.photos')
        var showIndexsInfo = getShowIndexs($container);

        var zIndex = 9999;
        var scales;
        var config = $container.data('config');
        var halfShowNum = (config.num - 1) / 2;
        var listWidth = halfShowNum * config.distance * 2 + config.maxWidth;
        var containerWidth = $container.width();
        var containerHeight = $container.height();
        if (containerWidth < listWidth) {
            $container.width(listWidth);
        }
        if (containerHeight < config.maxHeight) {
            $container.height(config.maxHeight);
        }
        $container.find('ul li img').css({
            width: "100%",
            height: "100%"
        });
        $container.find('ul').css({
            position: 'relative',
            width: listWidth,
            height: config.maxHeight,
            'list-style': 'none',
            padding: 0,
            margin: 0,
            marginLeft: '50%',
            marginRight: '50%',
            left: -listWidth / 2,
            /*customize:
            top: ($container.height() - config.maxHeight) / 2
            */
        });

        $container.find('.left').css({
            position: 'absolute',
            left: 10,
            top: '50%',
            'z-index': 9999 + $container.data('totalNum') + 1
        });

        $container.find('.right').css({
            position: 'absolute',
            right: 10,
            top: '50%',
            'z-index': 9999 + $container.data('totalNum') + 1
        });

        $container.find('ul li').each(function (index, element) {
            var i = showIndexsInfo.hashIndexs[index];
            if (i !== undefined) {
                scales = Math.pow(config.scale, Math.abs(i - halfShowNum));
                zIndex = 9999 + (i > halfShowNum ? (config.num - 1 - i) : i);
                $(element).css({
                    display: 'block',
                    position: 'absolute',
                    'z-index': zIndex,
                    overflow: 'hidden',
                    width: scales * config.maxWidth,
                    height: scales * config.maxHeight,
                    left: i * config.distance + (1 - scales) * config.maxWidth * Number(i > halfShowNum),
                    top: (1 - scales) * config.maxHeight / 2
                });
            } else {
                scales = Math.pow(config.scale, halfShowNum);
                $(element).css({
                    display: 'none',
                    position: 'absolute',
                    overflow: 'hidden',
                    width: scales * config.maxWidth,
                    height: scales * config.maxHeight,
                    left: halfShowNum * config.distance + (1 - scales) * config.maxWidth / 2,
                    top: 0
                });
            }

        });
    }


    $.fn.carousel = function (param) {
      // var $container = $('.photos')
        var config;
        var totalNum;
        var $target;
        $(this).each(function(index, target) {
            $target = $(target);
            if (typeof param === 'object' && param !== null) {
                config = getConfig(param);
                totalNum = $target.find('ul li').length;
                if (totalNum <= 0 || totalNum % 2 === 0) {
                    return;
                }
                if (config.num <= 0 || config.num > totalNum) {
                    config.num = totalNum;
                }
                $target.data('config', config);
                $target.data('index', 0);
                $target.data('totalNum', totalNum);
                initStyle($target);

                $target.find('.left').off('click').on('click', (function($target) {
                    return function() {
                        scroll($target, 'left');
                    }
                })($target));
                $target.find('.right').off('click').on('click',(function($target) {
                    return function() {
                        scroll($target, 'right');
                    }
                })($target));

                (function($target) {
                    var autoPlay;
                    clearInterval($target.data('auto'));
                    if($target.data('config').autoPlay) {
                        autoPlay = setInterval(function() {
                            scroll($target, 'right');
                        }, $target.data('config').showTime);
                        $target.data('auto', autoPlay);
                        $target.find('ul').off('mouseenter').on('mouseenter', function() {
                            clearInterval($target.data('auto'));
                        }).off('mouseleave').on('mouseleave', function() {
                            autoPlay = setInterval(function() {
                                scroll($target, 'right');
                            }, $target.data('config').showTime);
                            $target.data('auto', autoPlay);
                        });
                    } else {
                        $target.find('ul').off('mouseenter').off('mouseleave');
                    }
                })($target);
            }

        });

    };

})(jQuery);

